sdk/python/lib/pulumi/provider/experimental/ComponentProvider.md

```diff
@@ -0,0 +1,185 @@
+# Component Provider
+
+The `pulumi.provider.experimental` package allows writing a Pulumi provider in Python with minimal boilerplate, using type annotations to define the schema. The provider can be published to Git or referenced locally.
+
+## Building a Component Provider
+
+Create a new directory that will hold the provider implementation.
+
+First we need to let Pulumi's plugin system know which language runtime should be used to run the provider plugin. Create a a `PulumiPlugin.yaml` file with the contents:
+
+```yaml
+runtime: python
+```
+
+> [!NOTE]
+> Python component providers only support the pip toolchain with a virtual environment named `venv`. Don't add any other runtime options to the `PulumiPlugin.yaml` file.
+
+We also need to create a `requirements.txt` file to list the dependencies of the provider:
+
+```txt
+git+ssh://git@github.com/pulumi/pulumi.git@julienp/tutorial#subdirectory=sdk/python
+```
+
+TODO: update version to a pre-release with the experimental component provider merged
+
+
+The resources provided by the provider are are defined as Python classes that subclass `pulumi.ComponentResource`. Create a new file `my-component.py` to hold the component implementation:
+
+```python
+from typing import Optional
+
+import pulumi
+
+class MyComponent(pulumi.ComponentResource):
+    def __init__(self, name: str, opts: Optional[pulumi.ResourceOptions] = None):
+        super().__init__("my-provider:index:MyComponent", name, {}, opts)
+        self.register_outputs({})
+```
+
+> [!NOTE]
+> The type string `my-provider:index:MyComponent` follows the pattern `<package>:<module>:<type>`. The package is your provider's name, module must be `index`. The `type` must match the class name.
+
+Next, create a `__main__.py` file to host the provider:
+
+```python
+from pulumi.provider.experimental import Metadata, component_provider_host
+
+if __name__ == "__main__":
+    component_provider_host(
+        Metadata(name="my-provider", version="1.2.3", display_name="My Component Provider")
+    )
+```
+
+> [!NOTE]
+> The name passed to `Metadata` must exactly match the package name in the type string of your component.
+
+The call to `component_provider_host` will start the provider and listen for RPC requests from the Pulumi engine. Any subclasses of `pulumi.ComponentResource` found in any Python source code files in the provider's directory will be exposed as resources by the provider.
+
+At this point, we can run the provider, but it does not yet do anything useful. To get data in and out of the provider we need to define the inputs and outputs of the component. Inputs are infered from the `args` parameter of the component's constructor. Add a new type `MyComponentArgs` derived from `TypedDict` to represent the arguments, and add it as an argument to the constructor.
+
+```diff
+-from typing import Optional
++from typing import Optional, TypedDict
+
+ import pulumi
+
++class MyComponentArgs(TypedDict):
```

Comment:
Add docstrings and describe how they are translated to schema?

Reply:
hah, that reminds me that docstrings aren't implemented yet https://github.com/pulumi/pulumi/issues/18438